pkg/controller/operators/adoption_controller.go

```diff
@@ -0,0 +1,438 @@
+	if err != nil {
+		return err
+	}
+
+	enqueueCSV := &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(r.mapToClusterServiceVersions),
+	}
+	err = ctrl.NewControllerManagedBy(mgr).
+		For(&operatorsv1alpha1.ClusterServiceVersion{}).
+		Watches(&source.Kind{Type: &appsv1.Deployment{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.Namespace{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.Service{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.ServiceAccount{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.Secret{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.ConfigMap{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.Role{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.RoleBinding{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.ClusterRole{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.ClusterRoleBinding{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &apiextensionsv1.CustomResourceDefinition{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &apiregistrationv1.APIService{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &operatorsv1alpha1.Subscription{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &operatorsv1alpha1.InstallPlan{}}, enqueueCSV).
+		Complete(reconcile.Func(r.ReconcileClusterServiceVersion))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewAdoptionReconciler constructs and returns an AdoptionReconciler.
+// As a side effect, the given scheme has operator discovery types added to it.
+func NewAdoptionReconciler(cli client.Client, log logr.Logger, scheme *runtime.Scheme) (*AdoptionReconciler, error) {
+	// Add watched types to scheme.
+	if err := AddToScheme(scheme); err != nil {
+		return nil, err
+	}
+
+	factory, err := decorators.NewSchemedOperatorFactory(scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AdoptionReconciler{
+		Client: cli,
+
+		log:     log,
+		factory: factory,
+	}, nil
+}
+
+var fieldOwner = client.FieldOwner("olm")
+
+// ReconcileSubscription labels the CSVs installed by a Subscription as components of an operator named after the subscribed package and install namespace.
+func (r *AdoptionReconciler) ReconcileSubscription(req ctrl.Request) (reconcile.Result, error) {
+	// Set up a convenient log object so we don't have to type request over and over again
+	log := r.log.WithValues("request", req)
+	log.V(1).Info("reconciling subscription")
+
+	// Fetch the Subscription from the cache
+	ctx := context.TODO()
+	in := &operatorsv1alpha1.Subscription{}
+	if err := r.Get(ctx, req.NamespacedName, in); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Could not find Subscription")
+		} else {
+			log.Error(err, "Error finding Subscription")
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// OLM generated Operators are named after their packages and further qualified by the install namespace
+	if in.Spec == nil || in.Spec.Package == "" {
+		log.Info("subscription missing status, ignoring")
```

Comment:
nit: log message

Reply:
you mean to drop this?

Reply:
it says "subscription missing status" but the condition is "spec missing package"